library/src/main/java/com/novoda/downloadmanager/ServiceNotificationDispatcher.java

```diff
@@ -0,0 +1,84 @@
+                                  NotificationCreator<T> notificationCreator,
+                                  NotificationManager notificationManager) {
+        this.waitForDownloadService = waitForDownloadService;
+        this.notificationCreator = notificationCreator;
+        this.notificationManager = notificationManager;
+    }
+
+    @WorkerThread
+    void updateNotification(T payload) {
+        Wait.<Void>waitFor(service, waitForDownloadService)
+                .thenPerform(executeUpdateNotification(payload));
+    }
+
+    private Wait.ThenPerform.Action<Void> executeUpdateNotification(T payload) {
+        return () -> {
+            NotificationInformation notificationInformation = notificationCreator.createNotification(payload);
+
+            dismissStackedNotification(notificationInformation);
+
+            switch (notificationInformation.notificationStackState()) {
+                case SINGLE_PERSISTENT_NOTIFICATION:
+                    updateNotification(notificationInformation);
+                    break;
+                case STACK_NOTIFICATION_NOT_DISMISSIBLE:
+                    stackNotificationNotDismissible(notificationInformation);
+                    break;
+                case STACK_NOTIFICATION_DISMISSIBLE:
+                    stackNotification(notificationInformation);
+                    break;
+                default:
+                    String message = String.format(
+                            "%s: %s is not supported.",
+                            NotificationCustomizer.NotificationStackState.class.getSimpleName(),
+                            notificationInformation.notificationStackState()
+                    );
+                    throw new IllegalArgumentException(message);
+            }
+
+            return null;
+        };
+    }
+
+    private void updateNotification(NotificationInformation notificationInformation) {
+        service.startForeground(notificationInformation.getId(), notificationInformation.getNotification());
+    }
+
+    private void stackNotification(NotificationInformation notificationInformation) {
+        service.stopForeground(true);
+        Notification notification = notificationInformation.getNotification();
+        notificationManager.notify(NOTIFICATION_TAG, notificationInformation.getId(), notification);
+    }
+
+    private void stackNotificationNotDismissible(NotificationInformation notificationInformation) {
+        service.stopForeground(true);
+        Notification notification = notificationInformation.getNotification();
+        notification.flags |= Notification.FLAG_ONGOING_EVENT;
+        notificationManager.notify(NOTIFICATION_TAG, notificationInformation.getId(), notification);
+    }
+
+    private void dismissStackedNotification(NotificationInformation notificationInformation) {
+        notificationManager.cancel(NOTIFICATION_TAG, notificationInformation.getId());
+    }
+
+    void setService(Service service) {
```

Comment:
We could create a new interface rather than relying on the Android `Service` scope.

Reply:
Yes definitely... as you mentioned in other comments, but here is pretty evident

Reply:
79e4494